### DISCUSSION TITLE
What is the correct way of using your code with Homespan that doesn't lead to device freezing?

### DISCUSSION THREAD

**Discussion by AAfetisov** [HomeSpan/HomeSpan#913] [Q&A]
I added irremote library to listen to ir commands and to issue them, but then my device has never lasted even a day without freezing in different ways. It could be on wifi, but not responding to pings. It could be on homekit but there would be no incoming commands on serial. It could just wait for you to hit `enter` in serial to simultaneously launch every command you pressed in homekit in the last 3 minutes.
 Both libraries seem to work well apart from each other, never the less. Am I doing something wrong? I just fire ir commands when there are updates from homekit. I also listen to ir in the main loop, and then update homekit state with setVals ( something like currentTemp->setVal()). Thats all

**Comment by HomeSpan** [comment]
Depending on the library, listening for IR command may freeze HomeSpan (or any other background process).  HomeKit is ridiculously sensitive to timing and connectivity so anything that interrupts commands can cause a freeze.  However, it sounds like you the Serial monitor itself is also freezing.   This is likely a hangup with listening for IR commands.   What library are you using?

**Reply by AAfetisov** [reply]
thank you for the reply. Im using IRremote library. Took the code from their example

**Reply by HomeSpan** [reply]
Can you provide a link to the library?

**Reply by AAfetisov** [reply]
I think its this one https://github.com/Arduino-IRremote/Arduino-IRremote
im using the one from arduino ide
![2024-09-06-18 33@2x](https://github.com/user-attachments/assets/80eaa554-00dc-4d4d-8647-d3dff8ed88d0)

**Comment by HomeSpan** [comment]
I reviewed the library code and don't think it will work with HomeSpan.  As noted in the IRRemote README, this library has various contingencies, makes use of timers, interrupts, etc. that will likely interfere with other libraries, such as HomeSpan.

The ESP32 does have a built-in peripheral designing explicitly for sending IR commands.  I implemented this peripheral in HomeSpan a while back --- it's called the RFControl class.  Very robust and fully integrated with other HomeSpan functionality.   You may want to give it a try - see the RFControl docs link from the Main page of HomeSpan.  Also includes a variety of examples.  Unfortunately you will need to write your own code to generate the actual bytes codes to be sent.  I don't have anything that will help with that.

**Reply by AAfetisov** [reply]
I thought rtos could take care of timers etc. Oh well, I’ll try putting Irremote on some separate chip then
